File: leetcode/python3/word-search.py
#
# @lc app=leetcode id=79 lang=python
#
# [79] Word Search
#
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        is_visited = [[False for _ in range(len(board[0]))] for _ in range(len(board))]
        def dfs(grid, i, j, w):
            if not w:
                return True
            if not 0 <= i < len(grid) or not 0 <= j < len(grid[0]) or grid[i][j] != w[0]:
                return False
            if is_visited[i][j]:
                return False
            is_visited[i][j] = True
            res = dfs(grid, i+1, j, w[1:]) or dfs(grid, i-1, j, w[1:]) or dfs(grid, i, j+1, w[1:]) or dfs(grid, i, j-1, w[1:])
            is_visited[i][j] = False
            return res
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == word[0]:
                    if dfs(board, i, j, word):
                        return True
        return False        

File: leetcode/cpp/word-search.cpp
/*
 * @lc app=leetcode id=79 lang=cpp
 *
 * [79] Word Search
 */

// @lc code=start
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        for (int i = 0; i < board.size(); i++)
            for (int j = 0; j < board[i].size(); j++) {
                if (dfs(board, i, j, 0, word))
                    return true;
            }
        return false;
    }

    bool dfs(vector<vector<char>>& board, int i, int j, int u, string& word) {
        if (i < 0 || i >= board.size() || j < 0 || j >= board[i].size() || board[i][j] == ' ' || u >= word.size() || board[i][j] != word[u])
            return false;
        
        if (u == word.size()-1) return true;
        char c = board[i][j];
        board[i][j] = ' ';
        if (dfs(board, i+1, j, u+1, word) || dfs(board, i-1, j, u+1, word) ||
            dfs(board, i, j+1, u+1, word) || dfs(board, i, j-1, u+1, word))
            return true;
        board[i][j] = c;
        return false;
    }
};
// @lc code=end

